sqlalchemy-stubs/orm/query.pyi

```diff
@@ -45,9 +49,9 @@ class Query(object):
     def with_for_update(self, read: bool = ..., nowait: bool = ..., of: Optional[Any] = ...,
                         skip_locked: bool = ..., key_share: bool = ...): ...
     def params(self, *args, **kwargs): ...
-    def filter(self, *criterion): ...
-    def filter_by(self, **kwargs): ...
-    def order_by(self, *criterion): ...
+    def filter(self, *criterion) -> Query[T]: ...
+    def filter_by(self, **kwargs) -> Query[T]: ...
+    def order_by(self, *criterion) -> Query[T]: ...
```

Comment:
I think for these three and `limit` below it is better to use self-types, like:
```python
_Q = TypeVar('_Q', bound=Query[Any])

class Query(...):
    ...
    def filter(self: _Q, *criterion) -> _Q: ...
```

Reply:
Is there a semantic difference? :thinking:

Reply:
I makes a difference if someone subclasses `Query`. My understanding is that subclasses would return an instance of themselves.

Reply:
Ah yes, of course! Thanks for taking the time to enlighten me!